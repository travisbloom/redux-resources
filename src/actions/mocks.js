import { mockStore, normalizer, RESOURCE_REDUCER, RESOURCE, formatErrors } from '../mocks'
import createResource from './createResource'

const generateCreateResourceActions = async ({ request }) => {
    const store = mockStore({})
    const { createResources } = createResource({
        resource: RESOURCE_REDUCER,
        normalizer,
        request,
        formatErrors,
    })
    await store.dispatch(createResources(RESOURCE))
    return store.getActions()
}

module.exports = {
    generateCreateResourceActions,
}
