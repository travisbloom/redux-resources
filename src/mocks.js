import configureMockStore from 'redux-mock-store'
import thunk from 'redux-thunk'
import { Schema, normalize, arrayOf } from 'normalizr'

const RESOURCE_REDUCER = 'resources'

const RESOURCE_ID = 'resourceId'
const RESOURCE = {
    id: RESOURCE_ID,
    prop1: 'initial prop1',
    prop2: 'initial prop2',
}

const RESOURCE_UPDATE_PAYLOAD = {
    id: RESOURCE_ID,
    prop1: 'updated prop1',
    prop2: 'updated prop1',
}

const SERVER_ERRORS = [
    'This is an initial error',
    'This is another error',
]

const resourceSchema = new Schema(RESOURCE_REDUCER)
const normalizer = (resources) => {
    const array = Array.isArray(resources) ? resources : [resources]
    return normalize(array, arrayOf(resourceSchema))
}
const formatErrors = errors => errors

const middlewares = [thunk]
const mockStore = configureMockStore(middlewares)

module.exports = {
    RESOURCE_REDUCER,
    RESOURCE_ID,
    RESOURCE,

    RESOURCE_UPDATE_PAYLOAD,

    SERVER_ERRORS,

    mockStore,
    formatErrors,
    normalizer,
}
