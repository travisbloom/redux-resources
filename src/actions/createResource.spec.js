import { RESOURCE, SERVER_ERRORS } from '../mocks'
import { generateCreateResourceActions } from './mocks'

const request = () => Promise.resolve(RESOURCE)
const errorRequest = () => Promise.reject(SERVER_ERRORS)

it('will dispatch two actions on success', async () => {
    const actions = await generateCreateResourceActions({ request })
    expect(actions.length).toBe(2)
})

it('will dispatch two actions on error', async () => {
    const actions = await generateCreateResourceActions({ request: errorRequest })
    expect(actions.length).toBe(2)
})

it('will have an initial action', async () => {
    const [inititalAction] = await generateCreateResourceActions({ request })
    expect(inititalAction).toMatchSnapshot()
})

it('will have a success action', async () => {
    const [_, successAction] = await generateCreateResourceActions({ request })
    expect(successAction).toMatchSnapshot()
})

it('will have a error action', async () => {
    const [_, errorAction] = await generateCreateResourceActions({ request: errorRequest })
    expect(errorAction).toMatchSnapshot()
})
