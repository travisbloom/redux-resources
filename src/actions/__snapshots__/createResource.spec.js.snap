exports[`test will have a error action 1`] = `
Object {
  "meta": Object {
    "reduxResourcesActionType": "CREATE_RESOURCE_REQUEST_ERROR",
    "resource": "resources",
    "timestamp": "2016-01-01T05:00:00.000Z",
  },
  "payload": Array [
    "This is an initial error",
    "This is another error",
  ],
  "type": "CREATE_RESOURCES_REQUEST_ERROR",
}
`;

exports[`test will have a success action 1`] = `
Object {
  "meta": Object {
    "normalizedResponse": Object {
      "entities": Object {
        "resources": Object {
          "resourceId": Object {
            "id": "resourceId",
            "prop1": "initial prop1",
            "prop2": "initial prop2",
          },
        },
      },
      "result": Array [
        "resourceId",
      ],
    },
    "reduxResourcesActionType": "CREATE_RESOURCE_REQUEST_SUCCESS",
    "resource": "resources",
    "timestamp": "2016-01-01T05:00:00.000Z",
  },
  "payload": Object {
    "id": "resourceId",
    "prop1": "initial prop1",
    "prop2": "initial prop2",
  },
  "type": "CREATE_RESOURCES_REQUEST_SUCCESS",
}
`;

exports[`test will have an initial action 1`] = `
Object {
  "meta": Object {
    "reduxResourcesActionType": "CREATE_RESOURCE_REQUEST",
    "resource": "resources",
    "timestamp": "2016-01-01T05:00:00.000Z",
  },
  "payload": Object {
    "id": "resourceId",
    "prop1": "initial prop1",
    "prop2": "initial prop2",
  },
  "type": "CREATE_RESOURCES_REQUEST",
}
`;
